pkg/policy/rule.go

```diff
@@ -35,54 +35,93 @@ func (r *rule) String() string {
+// being merged has conflicting L7 rules with those already in the provided
+// L4PolicyMap for the specified port-protocol tuple, it returns an error.
 func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r api.PortRule, p api.PortProtocol,
 	proto api.L4Proto, ruleLabels labels.LabelArray, resMap L4PolicyMap) (int, error) {
 
 	key := p.Port + "/" + string(proto)
-	filter, ok := resMap[key]
+	existingFilter, ok := resMap[key]
 	if !ok {
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
-	l4Filter := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
-	if l4Filter.L7Parser != "" {
-		if filter.L7Parser == "" {
-			filter.L7Parser = l4Filter.L7Parser
-		} else if l4Filter.L7Parser != filter.L7Parser {
-			ctx.PolicyTrace("   Merge conflict: mismatching parsers %s/%s\n", l4Filter.L7Parser, filter.L7Parser)
-			return 0, fmt.Errorf("Cannot merge conflicting L7 parsers (%s/%s)", l4Filter.L7Parser, filter.L7Parser)
+
+	// Create a new L4Filter based off of the arguments provided to this function
+	// for merging with the filter which is already in the policy map.
+	filterToMerge := CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
+
+	// Handle cases where filter we are merging new rule with, or new rule itself
+	// allows all traffic on L3.
+	//
+	// Existing filter selects all endpoints, so don't add new endpoints.
+	if len(existingFilter.Endpoints) == 0 && len(filterToMerge.Endpoints) > 0 {
+		log.WithFields(logrus.Fields{
+			logfields.EndpointSelector: filterToMerge.Endpoints,
+			"policy":                   existingFilter,
+		}).Debug("skipping L4 filter as the endpoints are already covered")
+
+		// Existing L4Filter already selects all endpoints; if there are no L7 rules
+		// to add from the new rule, then we can just exit because the rule is as
+		// permissive as possible at L3, the L4 information is already contained
+		// within the filter, and there is no L7 metadata to add to the filter.
+		if r.NumRules() == 0 {
+			existingFilter.DerivedFromRules = append(existingFilter.DerivedFromRules, ruleLabels)
+			return 1, nil
+		}
+	} else {
+		// If new rule allows all endpoints, then allow all endpoints.
+		if len(existingFilter.Endpoints) > 0 && len(filterToMerge.Endpoints) == 0 {
+			log.WithFields(logrus.Fields{
+				logfields.EndpointSelector: filterToMerge.Endpoints,
+				"policy":                   existingFilter,
+			}).Debug("new L4 filter applies to all endpoints, making the policy more permissive")
+
+			// Allow all endpoints.
+			existingFilter.Endpoints = nil
+
+			// If new rule allows all endpoints and does have L7 rules, update
+			// filter's L7Parser.
+			if filterToMerge.L7Parser != ParserTypeNone && existingFilter.L7Parser == ParserTypeNone {
+				existingFilter.L7Parser = filterToMerge.L7Parser
+			}
 		}
+		existingFilter.Endpoints = append(existingFilter.Endpoints, endpoints...)
```

Comment:
The `append()` reverses the `existingFilter.Endpoints = nil` above. I think this should be in a else branch.

Reply:
This was actually the logic in `addEndpoints`, which was removed in eb8427a9c3942c8c1a556efdbcbb2f66916a7599 . Per the documentation for `Endpoints` in `L4Filter`: 

https://github.com/cilium/cilium/blob/9571490cbb45bfb93fa230a2c7a6101b03101a56/pkg/policy/l4.go#L88-L90

an empty slice represents all endpoints being selected.

I can test locally and see if a nil slice of EndpointSelector behaves the same as an empty slice.

Reply:
The code will work, it just looks a bit weird to me to set endpoints to nil and then set it to an empty slice again by appending it. I had to read the code twice to make sure. The append is only valid because endpoints is guaranteed to always be an empty slice. I'm OK as-is if you prefer it this way.

Reply:
Changed because your way makes it more clear; I'm all for more understandable code.